### PR TITLE
Lineigensolver and DataType

### PR DESCRIPTION
**Pull request by eringh**
Did a re-work of the eig-solvers, and some of the lin-solvers; see #95. Also worked on the `DataType` vs `Type`; see #139.

There are some questions remaining in #95, but this is approaching ready. Although I did not really manage to get type-stability, I believe the design is still favorable compared to the old one. At least for `NativeEigSolver` we know where the type-stability breaks, and for `NativeEigSSolver` I get that `eigs()` does not seem to be type stable (and hence I see no hope for us).

### CODE REVIEW

File: test/linsolver.jl [hunk 1]
```diff
@@ -90,4 +90,74 @@ using LinearAlgebra
...
+
+        for i = 1:3
+            @test λ1[i] ≈ λ2[i]
+        end
+
+        @info "NativeEigSSolver"
+        Random.seed!(0)
+        n = 20
+        A = sprand(ComplexF64, n, n, 0.25) + sparse(1.0*I,n,n)
+        B = sprand(ComplexF64, n, n, 0.25) + sparse(1.0*I,n,n)
```
Comment (maxbennedich) [inline test/linsolver.jl]: You can change `sparse(1.0*I,n,n)` to just `I`.

Comment (eringh) [reply test/linsolver.jl]: Good point! I changed it.
(I also did separate tests of all parts of increase encapsulation of the tests)

**Comment by eringh** [comment]
Now I start to feel ready with this work. I did not really manage to get type stability, but I have some problems with the type stability of the underlying functions. Perhaps I am calling them in an inappropriate way, or they are just unstable. It seems rather stable, except in terms of `Float64`/`ComplexF64`,for the generalized and dense problem.

**Comment by maxbennedich** [comment]
Looks like a nice cleanup to me!

**Comment by jarlebring** [comment]
:+1:

**Pull request merged**